misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -0,0 +1,192 @@
+
+internal class RealEventRouter : EventRouter {
+  @Inject lateinit var eventChannel: ClusterConnector
+  @Inject lateinit var eventJsonAdapter : JsonAdapter<SocketEvent>
+  @Inject @ForEventRouterSubscribers lateinit var executor: ExecutorService
+
+  lateinit var clusterSnapshot: ClusterSnapshot
+  private val subscriptionQueue = ArrayDeque<QueuedSubscribe<*>>()
+  private val publishQueue = ArrayDeque<QueuedPublish>()
+
+  // TODO: concurrent access to this stuff?
+  private val localSubscribers = LinkedHashMultimap.create<String, LocalSubscription<*>>()
+  private val remoteSubscribers = LinkedHashMultimap.create<String, WebSocket>()
+  private val hostToSockets = CacheBuilder.newBuilder()
+      .build<String, WebSocket>(object : CacheLoader<String, WebSocket>() {
+        override fun load(topicOwner: String): WebSocket {
+          return eventChannel.connectSocket(topicOwner, webSocketListener)
+        }
+      })
+
+  val webSocketListener = object : WebSocketListener() {
+    override fun onMessage(webSocket: WebSocket, text: String) {
+      val event = eventJsonAdapter.fromJson(text)!!
+      val (topic, message) = event
+      if (message == "subscribe") {
+        remoteSubscribers.put(topic, webSocket)
+        return
+      }
+
+      // TODO: unsubscribe
+      // TODO: type safety or something for the string "subscribe"
+
+      localSubscribers.get(topic).forEach {
+        it.enqueue(executor, LocalSubscription.Action.Event(message))
+      }
+
+      remoteSubscribers.get(topic).forEach {
+        it.send(eventJsonAdapter.toJson(event))
+      }
+    }
+  }
+
+  private val topicPeer = object : TopicPeer {
+    override fun acceptWebSocket(webSocket: WebSocket): WebSocketListener = webSocketListener
+
+    override fun clusterChanged(clusterSnapshot: ClusterSnapshot) {
+      this@RealEventRouter.clusterSnapshot = clusterSnapshot
+      initCluster()
+    }
+  }
+
+  /** We queue up subscribes when we don't have a cluster. */
+  private fun initCluster() {
+    while (subscriptionQueue.isNotEmpty()) {
+      val queuedSubscribe = subscriptionQueue.pop()
+      subscribe(queuedSubscribe)
+    }
+
+    while (publishQueue.isNotEmpty()) {
+      val (topic, event) = publishQueue.pop()
+      publish(topic, event)
+    }
+  }
+
+  fun joinCluster() {
+    eventChannel.joinCluster(topicPeer)
+  }
+
+  fun leaveCluster() {
+    eventChannel.leaveCluster(topicPeer)
+  }
+
+  override fun <T : Any> getTopic(name: String): Topic<T> {
+    return object : Topic<T> {
+      override val name: String
+        get() = name
+
+      override fun publish(event: T) {
+        this@RealEventRouter.publish(name, event)
+      }
+
+      override fun subscribe(listener: Listener<T>): Subscription<T> {
+        return this@RealEventRouter.subscribe(this, listener)
+      }
+    }
+  }
+
+  private fun <T : Any> publish(topic: String, event: T) {
+    if (!::clusterSnapshot.isInitialized) {
+      publishQueue.add(QueuedPublish(topic, event))
```

Comment:
I’d like this to _always_ do a QueuedPublish, and then call some helper method to drain the queue